### PR TITLE
Installs the .la files in the install location

### PR DESCRIPTION
**Pull request by permcody**
This PR appears to work on Mac for installed MOOSE applications. However, there are as many unanswered questions as there are answered ones with this. First to get this working for our dynamic load case (MacOS only), one most set `DYLD_LIBRARY_PATH` to point to the multiapp's lib location. This is probably not something we want to do in the long run. I have no idea how this behaves yet on Linux. 

What I'm discovering though is that Mac may be automatically loading dependent libraries. I'm making this assumption based on the following error:
```
dlopen(/Users/permcj/projects/installed/lib/bison/libxfem-opt.0.dylib, 0x0001): Library not loaded: @rpath/libtensor_mechanics-opt.0.dylib
  Referenced from: <2DFB3AA0-1CE7-3564-A3FF-96F2369FCAC0> /Users/permcj/projects/installed/lib/bison/libxfem-opt.0.dylib
  Reason: tried: '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/lib/pronghorn/./libtensor_mechanics-opt.0.dylib' (no such file), '/usr/local/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/usr/lib/libtensor_mechanics-opt.0.dylib' (no such file, not in dyld cache)
```

Note all of the paths being search - automatically. This is the second library that's "loaded" by MOOSE. The first being the one being searched for above.

**UPDATE (5/1):** With the latest patch this appears to be working on both Mac and Linux with in-tree and installed binaries. Also the double-free error on tear down is gone!

There were two major confounding factors that lead to what I believe was an implementation problem of this system _since its inception_. Namely, it doesn't appear that manually loading "inter-library dependencies" is necessary despite my long held belief. I believe that the primary issue was the lack of uniqueness in a specific symbol used for loading that was introduced to create our beautiful modules system. Namely the `Modules::registerAll` method is duplicated for each application even though the implementation in the body of that method is dependent based on the specific set of modules enabled. While this design error was successfully covered by compiling applications together (since our library names are unique), it appears to have been a primary cause of some of the longstanding difficulty of missing module registration for our dynamically opened libraries. With the latest patches, I'm successfully opening _just_ the top-level application library dependency and successfully running input files that include objects that are _not_ in the driving application. 

Fixes: With this PR, I'm deprecating `Modules::registerAll` in favor of `Modules::registerAllObjects<AppName>`. This will effectively give us back a unique entry point for each application. I'm also turning off the manually recursive library load logic. 

Intermediate Debugging and Rollout: For right now I'm going to continue to install the .la files, but if everything in this PR works out as planned, we won't need those anymore. The libraries themselves should contain the information needed to properly load the dependent libraries with out library archive files so that will come back out. Also, I'm going to add a Boolean to allow us to toggle the manually recursive loading just in case. Again, I don't expect we'll need this, but it maintains the logic we've had in there for the past ~8 years at least for now in case we need it in a pinch. That can also come out after sufficient testing/usage. Apps (at least the nuclear herd) will need to update to the new module registration method, but the old one can stick around for a while. Several months down the road we should be able to significantlly clean all of this loading logic up. We can probably cut the total lines in those routines by upwards of 70% since we only need to load a single library now. 

Testing: We still don't have a good way of testing all of this. I know some want to create new dummy applications that emulate some of our more complex herd animals, but I don't think that's effective. I've found through all of the implementation that being close (like I was 8 years ago) is not anywhere close to good enough. We need to understand why a specific combination isn't working with all the real-world messy code that each of these applications brings along. At the very least we should be using real applications to test even if we pick specific input files to use in testing. For my testing, I've been using real Pronghorn and BISON input files and using `FullSolveMultiapp` to just run the subapp input file. That's generally enough to give me a warm fuzzy that the subapp is being properly dynamically loaded and fully working. Running `simple_transient_diffusion` is not good enough. We need to know that downstream dependencies are being used and working properly. With the BISON/Pronghorn example, I found many input files that worked because many were using a subset of objects that Pronghorn new about. It wasn't until I stumbled upon files containing XFEM (a module not used by Pronghorn) that I started to unravel this whole mystery. So for testing we do need disjoint subsets of modules enabled and input files that exercise objects from the set difference. 

Should we merge this before we figure out all the testing?
I'm biased, but yes. We know that the halfway baked implementation is already working well for some of our users (notably the microreactor team at ANL), but that's because they are getting lucky with modules. This PR installs the .la files (but those normally won't be used, which also means the python script to patch them also won't be used). Again, we'll pull all that back out later. The main "merge me now" changes here are the two I started this update with (adding a new templated modules registration method - very low impact, and turning off recursive loading). The former is a new method so won't affect existing testing and the second feature can only affect dynamic loading.

### CODE REVIEW

File: framework/build.mk
Comment (lindsayad) [inline framework/build.mk:16]: I feel like I've been destroyed by defaults like this in the past

Comment (lindsayad) [reply framework/build.mk:16]: But I guess this a default for where *you* are installing? Not a default for where to look for packages?

Comment (permcody) [reply framework/build.mk:16]: `/usr/local` is the normal "automake" default and yes, this is for where you will install, not look for anything. However, since we don't require that users run "configure". We evaluate $PREFIX" in our Makefile and it's empty. So our "no configure" default is actually `/`. If you run "configure" without any arguments, you get an explicit PREFIX of `/usr/local`. To me, it's sane to make those two the same, which this will do.

File: framework/app.mk
Comment (lindsayad) [inline framework/app.mk:438]:
```suggestion
```

File: framework/app.mk
Comment (lindsayad) [inline framework/app.mk:492]:
```suggestion
```

File: framework/scripts/patch_la.py
Comment (lindsayad) [inline framework/scripts/patch_la.py:39]: Should we error if the length is something other than 3?

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1881]:
```suggestion
```

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1883]:
```suggestion
  // implications for applications that have "diamond" inheritance of libraries (usually
```

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1884]:
```suggestion
  // modules). We will only load one of those libraries, versions be damned.
```

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1889]:
```suggestion
    const auto dl_lib_full_path = dir + '/' + dl_lib_filename;
```

File: framework/src/base/MooseApp.C
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1996]:
```suggestion
  if (!library_path.empty())
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/scripts/patch_la.py
Comment (permcody) [reply framework/scripts/patch_la.py:39]: Maybe, I'm following the exact same patter of our `get_repo_revision` script that we've used for more than a decade. It just doesn't do anything if it's not called exactly right, which is OK. I guess that other script just doesn't store information in the generated header. This script failing might actually be catastrophic.

**Comment by lindsayad** [comment]
So we really can't do better than the 3.85% coverage rate?

File: framework/src/multiapps/MultiApp.C
Comment (lindsayad) [inline framework/src/multiapps/MultiApp.C:85]:
```suggestion
                        "necessary and is hear for debugging/troubleshooting.");
```

File: modules/module_loader/include/ModulesApp.h
Comment (lindsayad) [inline modules/module_loader/include/ModulesApp.h:123]:
```suggestion
```

File: modules/module_loader/include/ModulesApp.h
Comment (lindsayad) [inline modules/module_loader/include/ModulesApp.h:229]:
```suggestion
  libmesh_ignore(f, s, af);
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by loganharbour** [comment]
> So we really can't do better than the 3.85% coverage rate?

I can try to add installed runs to coverage. But I really have _no_ idea how that'd work.

**Comment by permcody** [comment]
> > So we really can't do better than the 3.85% coverage rate?
> 
> I can try to add installed runs to coverage. But I really have _no_ idea how that'd work.

Looking at the coverage report now. So the `load_dependencies` branch is always false now (that is the piece that will be used for debugging only, but should be removable). The new `getLibrarySearchPaths` method only gets called on failure, which we admittedly don't have. The one thing that I don't understand is why some of the other lines aren't covered (like the structured binding). That's right in the base implementation. I'm guessing we only do coverage on the main module run and not on "test subs"? Note that the dynamic loading tests only run on the latter. 

Testing for this capability is bad no matter what in lieu of something new (like I described on the PR). If we want testing that actually matters, we're going to need to engineer some real tests with some of the nuclear herd. I have ideas on that. I'll start the dialog with Logan and Jason in the cube tomorrow. I would suggest that we don't hold this up too much though in lieu of testing. I don't have much assurance that any of the existing tests are good enough to test what we need to test.

**Comment by permcody** [comment]
In lieu of testing, my plan will be to work with Yinbin (on the Microreactor apps) and April (on Cardinal) as soon as this is merged. Yinbin is already working with dynamic loading and updating his apps to use it by default. Ideally, they can quite compling their own "mega app. April is trying to do the same. They are compiling Griffin into Cardinal, but she's very interested on using this capability. For her, the current code is failing so she's still compiling into Cardinal. With this PR merged, I want to give it another shot.

File: framework/include/base/MooseApp.h
Comment (permcody) [reply framework/include/base/MooseApp.h:975]: It strictly doesn't, but I plan to pull this out sooner than later. The whole recursive routine can hopefully go! I'm not kidding when I say I expect 70% of the existing lines to go. The new logic should be to load the named module given the library paths and fail when not found, period. No opening and parsing library archives, no recursive loading.

File: modules/module_loader/include/ModulesApp.h
Comment (permcody) [reply modules/module_loader/include/ModulesApp.h:123]: I don't understand why these are here. I didn't put them in. They are called in the current registration methods. They've always been empty. Maybe something Robert was doing?

File: modules/module_loader/include/ModulesApp.h
Comment (permcody) [reply modules/module_loader/include/ModulesApp.h:229]: Any thoughts on why these are here in the first place? It was a I don't understand what these are for so don't touch them, "inaction".

Comment (lindsayad) [reply modules/module_loader/include/ModulesApp.h:229]: If no modules are defined, then these variables would be unused

**Review by lindsayad — DISMISSED** [review]

**Review by lindsayad — APPROVED** [review]

**Comment by permcody** [comment]
Thanks Alex!

**Comment by loganharbour** [comment]
We're more or less ready to test this at scale with all of our codes with the container squashing.

**Comment by permcody** [comment]
We need to update all the apps (at least the ones we care about) to call
the new templated registration method. Then things should work!

**Pull request merged**
